### PR TITLE
8315966: Relativize initial_sp in interpreter frames

### PR DESCRIPTION
**Pull request by fbredber**
Relativize initial_sp in interpreter frames.

By changing the "initial_sp" (AKA "monitor_block_top" or "monitors" on PowerPC) member in interpreter frames from being an absolute address into an offset that is relative to the frame pointer, we don't need to change the value as we freeze and thaw frames of virtual threads. This is since we might freeze and thaw from and to different worker threads, so the absolute address to locals might change, but the offset from the frame pointer will be constant.

This subtask only handles relativization of "initial_sp" and "monitor_block_top" since it's the same slot in interpreter frames (roughly the same as "monitors" on PowerPC). Relativization of other interpreter frame members are handled in other subtasks to JDK-8289296.

Tested tier1-tier7 on supported platforms. The rest was sanity tested using Qemu.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8315966](https://bugs.openjdk.org/browse/JDK-8315966): Relativize initial_sp in interpreter frames (**Sub-task** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)
 * [Martin Doerr](https://openjdk.org/census#mdoerr) (@TheRealMDoerr - **Reviewer**) ⚠️ Review applies to [29b576fd](https://git.openjdk.org/jdk/pull/15815/files/29b576fde344dba0934e6abcf9ff81798d809a55)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**) ⚠️ Review applies to [dad25b66](https://git.openjdk.org/jdk/pull/15815/files/dad25b662c26dcf4261851e75240851fcd4a1a88)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15815/head:pull/15815` \
`$ git checkout pull/15815`

Update a local copy of the PR: \
`$ git checkout pull/15815` \
`$ git pull https://git.openjdk.org/jdk.git pull/15815/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15815`

View PR using the GUI difftool: \
`$ git pr show -t 15815`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15815.diff">https://git.openjdk.org/jdk/pull/15815.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15815#issuecomment-1725408887)

### CODE REVIEW

**Comment by TheRealMDoerr** [comment]
Thanks for implementing the platform parts! Note that you could have used R0 which is typically available as scratch reg. But, ok, we have enough other regs. Looks correct. I'll run more tests.

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp [hunk 1]
```diff
@@ -700,8 +700,10 @@ void InterpreterMacroAssembler::remove_activation(
     bind(restart);
     // We use c_rarg1 so that if we go slow path it will be the correct
     // register for unlock_object to pass to VM directly
-    ld(c_rarg1, monitor_block_top); // points to current entry, starting
-                                     // with top-most entry
+    ld(c_rarg1, monitor_block_top); // derelativize pointer
+    shadd(c_rarg1, c_rarg1, fp,  c_rarg1, LogBytesPerWord);
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/interp_masm_riscv.cpp:704]: Nit: One redundant space between the 3rd and 4th parameters for each `shadd` call added.

**Review by RealFYang — COMMENTED** [review]
Hi, I have arranged tier1-3 test on linux-riscv64 platform. Thanks for adding handling for riscv.

**Review by TheRealMDoerr — APPROVED** [review]
Tests have passed and the PPC64 parts look good.

**Comment by RealFYang** [comment]
> Hi, I have arranged tier1-3 test on linux-riscv64 platform. Thanks for adding handling for riscv.

Tier1-3 test is clean. The riscv part looks good except for the nit.

**Review by pchilano — APPROVED** [review]
x86 and AArch64 parts look good to me.

Thanks,
Patricio

File: src/hotspot/cpu/riscv/interp_masm_riscv.cpp (hunk 1 above)
Comment (fbredber) [reply src/hotspot/cpu/riscv/interp_masm_riscv.cpp]: Fixed

**Comment by fbredber** [comment]
Thanks for the review { @RealFYang, @pchilano, @TheRealMDoerr }.

The "Updated after review" version contains the following changes.

1. The extra space in `interp_masm_riscv.cpp` was removed.
2. I'm now using `R0` as a scratch register in `InterpreterMacroAssembler::save_interpreter_state` on PowerPC.
3. I removed some local variables in `FreezeBase::relativize_interpreted_frame_metadata` and `ThawBase::derelativize_interpreted_frame_metadata` that are no longer needed (x86, AArch64 and RISC-V).

Runs tier1-4 ok on supported platforms. PowerPC and RISC-V was sanity tested using Qemu.
Please check if still looks ok.

**Comment by RealFYang** [comment]
> The "Updated after review" version contains the following changes.
> 
> 1. The extra space in `interp_masm_riscv.cpp` was removed.

I see `shadd` added in other files are not covered? They have similar issues. Looks good otherwise. Thanks.

**Review by TheRealMDoerr — APPROVED** [review]
PPC64 part looks great! Thanks!

**Comment by fbredber** [comment]
Thank you all. If no one else has anything to add, I'll integrate (as soon as I can convince a sponsor).

**Pull request closed without merging**